Sources/_StringProcessing/Compiler.swift

```diff
@@ -17,30 +17,71 @@ struct RegexProgram {
 }
 
 class Compiler {
```

Comment:
I'm okay with merging the current implementation to unblock things, but the reason I made `Compiler` a class that owns an AST is because we could add environment properties to it later, e.g. function tables, cached `hasCaptures`, etc that are specific to the current AST being compiled.  As such it doesn't seem ideal to make `Compiler` stateless to accommodate two types of ASTs.

Reply:
Do you think you an sketch out what you were thinking a bit more?

Reply:
I haven't thought very carefully about this... Perhaps the `Compiler` here can act as the shared state that's generic over an AST-specific struct that actually performs code emission.

Reply:
Long-term, we'll have an optimizable IR (which is also our ABI) and lowering/emission would be a separate phase. I think it makes sense for phases to have their own state, and there might be some whole-program state. Typically I'd expect lowering or instruction selection to have self-contained state.

I think something interesting and under-explored-so-far is our caching strategy for run-time compilation. I know the `Regex` type has a program in it, but I don't think we've interwoven options (which affect compilation) into it yet.